### PR TITLE
[3.1] Re-implement SameSite for 2019

### PR DESCRIPTION
**Pull request by Tratcher**
@blowdart  #12125

Cherry picked from 2.1 (#13746) and updated with new enum values for Unspecified.

### CODE REVIEW

**Comment by blowdart** [comment]
Feels breaking. What's the effect on Cookie Policy here? Shouldn't the underlying cookie quirks be enough?

**Comment by Tratcher** [comment]
CookiePolicy also has to account for the new level of "None" vs "Unspecified". I didn't add quirks at the CookiePolicy layer because you're right, that's handled at the SetCookieHeaderValue layer.

**Comment by Tratcher** [comment]
Added more quirks for CookiePolicy, etc..

**Review by HaoK — APPROVED** [review]

**Comment by Tratcher** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=372045

**Comment by Tratcher** [comment]
@anurse this is ready to merge.

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/3173
https://github.com/aspnet/AspNetCore-Internal/issues/2742

File: src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs [hunk 1]
```diff
@@ -32,14 +32,36 @@ public Startup(IConfiguration config, IWebHostEnvironment env)
         public IConfiguration Configuration { get; set; }
         public IWebHostEnvironment Environment { get; }
 
+        private void CheckSameSite(HttpContext httpContext, CookieOptions options)
+        {
+            if (options.SameSite > SameSiteMode.Unspecified)
+            {
+                var userAgent = httpContext.Request.Headers["User-Agent"];
```
Comment (Tratcher) [inline src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs]: `var userAgent = httpContext.Request.Headers["User-Agent"].ToString();`

Comment (anurse) [reply src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs]: Do you need to make this change in the preview1 build?

Comment (Tratcher) [reply src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs]: No, it's just a sample, I've corrected it elsewhere. I just wanted to mark the typo so I'd remember it.

**Pull request merged**